bindings/container.cpp

```diff
@@ -13,28 +13,44 @@ Container::Container(QObject *parent) : QObject(parent)
 
 }
 
-void Container::installApp(QString path) {
-    if(!path.startsWith(":/")) path = QUrl(path).toLocalFile();
+Dna* loadDnaFile(QString path) {
+    if(!path.startsWith(":/"))
+        if(QUrl(path).toLocalFile() != "")
+            path = QUrl(path).toLocalFile();
     QFile file(path);
+
     if(file.open(QIODevice::ReadOnly | QIODevice::Text)) {
         QByteArray contents = file.readAll();
         char* buf = contents.data();
 
-        Dna *dna = holochain_dna_create_from_json(buf);
+        return holochain_dna_create_from_json(buf);
+    } else {
+        return 0;
+    }
+}
 
-        if(!dna) {
-            std::cout << "No valid Holochain DNA file" << std::endl;
-            std::cout << QString("%1 is not a valid Holochain DNA file!").arg(path).toStdString() << std::endl;
-            return;
-        }
+void Container::installApp(QString path) {
 
-        QCryptographicHash hash(QCryptographicHash::Sha256);
-        hash.addData(contents);
+    Dna *dna = loadDnaFile(path);
 
-        QSettings settings;
-        settings.beginGroup("dnas");
-        settings.setValue(hash.result().toBase64(), contents);
-        emit appsChanged();
+    if(!dna) {
+        std::cout << "No valid Holochain DNA file" << std::endl;
+        std::cout << QString("%1 is not a valid Holochain DNA file!").arg(path).toStdString() << std::endl;
+        return;
+    } else {
+        if(!path.startsWith(":/")) path = QUrl(path).toLocalFile();
+        QFile file(path);
+
+        if(file.open(QIODevice::ReadOnly | QIODevice::Text)) {
+            QByteArray contents = file.readAll();
```

Comment:
Just confirming that QFile read does an exact byte for byte read and doesn't try to interpret and covert things based on encoding type?  Because if it does, then the hash will be different...

Reply:
Yes, byte exact read which is why it returns a QByteArray.